src/Compiler/Opts/ConstantFold.idr

```diff
@@ -0,0 +1,164 @@
+module Compiler.Opts.ConstantFold
+
+import Compiler.CompileExpr
+import Core.Context
+import Core.Context.Log
+import Core.Primitives
+import Core.Value
+import Data.List
+import Data.Vect
+
+find : FC -> SizeOf outer -> CExp inner -> Var (outer ++ [x] ++ inner) -> CExp (outer ++ inner)
+find fc out val var = case sizedView out of
+    Z => case var of
+        MkVar {i=Z} First => val
+        MkVar {i=S _} (Later p) => CLocal fc p
+    S out' => case var of
+        MkVar {i=Z} First => CLocal fc First
+        MkVar {i=S _} (Later p) => weaken $ find fc out' val (MkVar p)
```

Comment:
This function should probably be reorganised to call `weakenN` once instead of
calling `weaken` many times (`SizeOf outer`). Something like:

```
aux : SizeOf outer -> CExp inner -> Var (outer ++ [x] ++ inner) ->
	  Either (Var (outer ++ inner)) (CExp inner)
```

should be doable.